### ISSUE TITLE
api.root.remove_product segfault when the product is a type with ifcmateriallayersetusage

### ISSUE THREAD

**Issue by brunopostle** [IfcOpenShell/IfcOpenShell#2697]
I can't seem to remove a type when it has a material layer set usage.

This seems to be a valid IFC file [_test2.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/10418906/_test2.txt) containing a single IfcWallType.

removing the type results in a segfault:

```
import ifcopenshell
from ifcopenshell.api import run
file = ifcopenshell.open("_test2.ifc")
walltype = file.by_type("IfcWallType")[0]
run("root.remove_product", file, product=walltype)

**Comment by aothms** [comment]
"The issue" appears to be that you have both an association to the MLS as well as the Usage.

```
#99=IFCRELASSOCIATESMATERIAL('3Ji8dCY2T0$OFJjapA5OnT',$,$,$,(#98),#104);
#131=IFCRELASSOCIATESMATERIAL('2IetrgebwergerdfsdNdQt',$,$,$,(#98),#132);
```

We're probably looping over a list of relationships or inverses and because removing one affects the other, the list being iterated over isn't valid anymore.

This could be hard to resolve :(

**Comment by brunopostle** [comment]
Looking at the spec, a Type isn't allowed to have a Usage, we are supposed to set the offset each time we create a wall 🙄, though this is not likely the cause of the segfault:

> NOTE  As a material layer set usage is an occurrence based information, that applies to each individual element, it cannot be assigned to an element type.

**Comment by Moult** [comment]
Thanks @aothms! Yes 3.8 is fine, in fact officially now I've made the leap and claimed that we only support 3.10 (as influenced by the last three Blender releases): https://blenderbim.org/docs/devs/system_support.html

I guess this can be closed, yes usages only appear on occurrences, this is one of those things I've proposed types to have to at least indicate a "default" usage. Another example of this inconvenience is whether the layerset is AXIS2 or AXIS3, should the type be wall-like or slab-like? In the case of coverings, it's ambiguous.
